mdal/mdal_memory_data_model.cpp

```diff
@@ -0,0 +1,209 @@
+/*
+ MDAL - Mesh Data Abstraction Library (MIT License)
+ Copyright (C) 2018 Peter Petrik (zilolv at gmail dot com)
+*/
+
+#include "mdal_memory_data_model.hpp"
+#include <assert.h>
+#include <math.h>
+#include <cstring>
+#include <algorithm>
+#include "mdal_utils.hpp"
+
+MDAL::MemoryDataset::~MemoryDataset() = default;
+
+size_t MDAL::MemoryDataset::activeData( size_t indexStart, size_t count, int *buffer )
+{
+  assert( parent );
+  if ( parent->isOnVertices() )
+  {
+    assert( active.size() > indexStart ); //checked in C API interface
+    assert( active.size() >= indexStart + count ); //checked in C API interface
+    int *src = active.data() + indexStart;
+    memcpy( buffer, src, count * sizeof( int ) );
+  }
+  else
+  {
+    memset( buffer, true, count * sizeof( int ) );
+  }
+
+  return count;
+}
+
+size_t MDAL::MemoryDataset::scalarData( size_t indexStart, size_t count, double *buffer )
+{
+  assert( parent ); //checked in C API interface
+  assert( parent->isScalar() ); //checked in C API interface
+
+  size_t i = 0;
+  while ( true )
+  {
+    if ( indexStart + i >= values.size() )
+      break;
+
+    if ( i >= count )
+      break;
+
+    const MDAL::Value value = values[ indexStart + i ];
+    if ( value.noData )
+    {
+      buffer[i] = MDAL_NAN;
+    }
+    else
+    {
+      buffer[i] = value.x;
+    }
```

Comment:
why do we need this `if (nodata)` test here? can't we just have "values" array filled up with NaN values where needed?